### PR TITLE
rePush class LinkList,implemented class LinkList and text

### PR DESCRIPTION
**Pull request by VivianWenhuiLiang**
implement class LinkedList and tested

### CODE REVIEW

File: LinkList.java [hunk 1]
```diff
@@ -1,40 +1,64 @@
+import java.util.NoSuchElementException;
+@SuppressWarnings("unchecked")
```
Comment (GeorgeZhiXu) [inline LinkList.java]: As discussed, add this on method level. Don’t suppress for the entire class.

File: LinkList.java [hunk 2]
```diff
@@ -43,11 +67,12 @@ public E get(int index) {
...
             temp=temp.next;
         }
-        return temp.data;
+        return (E)temp.data;
     }
 
+    @SuppressWarnings("unchecked")
     public E remove(int index) {
-        if(index<0||index<=size()) {
+        if(index<0||index>=size()) {
```
Comment (GeorgeZhiXu) [inline LinkList.java]: This boundary check is exactly the same as in that in “get” method. Make be convert them into a common method?

File: LinkList.java [hunk 3]
```diff
@@ -71,40 +97,27 @@ public boolean remove(Object o) {
...
         }
         Node temp=head;
         E holder;
         for(int i=0;i<size();i++) {
             if(i==index){
-                holder=temp.data;
+                holder=(E)temp.data;
                 temp.data=element;
-                return holder;
+                return (E)holder;
```
Comment (GeorgeZhiXu) [inline LinkList.java:111]: holder is already in type E, why do you need to cast it again?

File: LinkList.java [hunk 4]
```diff
@@ -120,34 +133,153 @@ public boolean contains(Object o) {
 
      }
 
-     public static void main(String [ ] args) {
-        ArrayList a=new LinkList();
+      public Iterator<E> iterator() {
+        return null;
```
Comment (GeorgeZhiXu) [inline LinkList.java:137]: Add a TODO

File: LinkList.java [hunk 5]
```diff
@@ -120,34 +133,153 @@ public boolean contains(Object o) {
...
+    }  
+
+    public void addFirst(E e) {
+        if(size()==0){
+            head.data=e;
+            numNodes++;
+            return;
+        }
```
Comment (GeorgeZhiXu) [inline LinkList.java:145]: Would it be easier if you create a newHead and then newHead.next = head; head = newHead;

File: LinkList.java [hunk 6]
```diff
@@ -120,34 +133,153 @@ public boolean contains(Object o) {
...
+            tra=tra.next;
+        }
+        Node temp=new Node<E>();
+        temp.data=e;
+        tra.next=temp;
+        numNodes++;
+    }
+
+    @SuppressWarnings("unchecked")
+    public E getFirst() {
```
Comment (GeorgeZhiXu) [inline LinkList.java:171]: Is it supposed to return null or throw exception? Please double check the doc.

File: LinkList.java [hunk 7]
```diff
@@ -120,34 +133,153 @@ public boolean contains(Object o) {
...
+        return true;
+    }
+
+    public E element() {//retrives the head of the list
+        return getFirst();
+    }
+
+
+    public E peek() {
+        if(isEmpty()) {
```
Comment (GeorgeZhiXu) [inline LinkList.java:233]: Good. This is better than if(size()==0) as the semantic is more clear

**Review by GeorgeZhiXu — CHANGES_REQUESTED** [review]

File: LinkList.java (hunk 1 above)
Comment (VivianWenhuiLiang) [reply LinkList.java]: will fixed later

File: LinkList.java (hunk 2 above)
Comment (VivianWenhuiLiang) [reply LinkList.java]: fixed

File: LinkList.java (hunk 3 above)
Comment (VivianWenhuiLiang) [reply LinkList.java:111]: ok,fixed

File: LinkList.java (hunk 4 above)
Comment (VivianWenhuiLiang) [reply LinkList.java:137]: added

File: LinkList.java (hunk 6 above)
Comment (VivianWenhuiLiang) [reply LinkList.java:171]: throw new NoSuchElementException(); Fixed

File: LinkList.java (hunk 7 above)
Comment (VivianWenhuiLiang) [reply LinkList.java:233]: ok

**Review by GeorgeZhiXu — APPROVED** [review]

**Review by GeorgeZhiXu — APPROVED** [review]

File: LinkList.java (hunk 5 above)
Comment (VivianWenhuiLiang) [reply LinkList.java:145]: good tip

**Review by GeorgeZhiXu — APPROVED** [review]
Let’s merge this and close the pull request.

**Pull request merged**